Aquarius/Src/Aquarius/Renderer/VertexArray.h

```diff
@@ -0,0 +1,120 @@
+#pragma once
+
+#include "Aquarius/Renderer/IndexBuffer.h"
+#include "Aquarius/Renderer/VertexBuffer.h"
+
+#include <cstdint>
+#include <glad/glad.h>
+#include <initializer_list>
+#include <memory>
+#include <unordered_map>
+#include <vector>
+
+
+namespace Aquarius {
+
+    enum class ShaderType
+    {
+        Int,
+        UInt,
+        Float,
+        Double,
+    };
+
+    static const std::unordered_map<ShaderType, GLenum> ShaderTypeToGLEnum =
+    {
+        {ShaderType::Int, GL_INT},
+        {ShaderType::UInt, GL_UNSIGNED_INT},
+        {ShaderType::Float, GL_FLOAT},
+        {ShaderType::Double, GL_DOUBLE}
+    };
+
+    static const std::unordered_map<ShaderType, uint32_t>  ShaderTypeSizeBytes =
+    {
+            {ShaderType::Int, 4},
+            {ShaderType::UInt, 4},
+            {ShaderType::Float, 4},
+            {ShaderType::Double, 8}
+    };
+
+    class VertexElement
+    {
+    public:
+        VertexElement(ShaderType shaderType, uint32_t count, bool normalize)
+        {
+            m_Type = ShaderTypeToGLEnum.at(shaderType);
+            m_Count = count;
+            m_Size = m_Count * ShaderTypeSizeBytes.at(shaderType);
+            m_Normalize = normalize;
+            m_Offset = 0;
+        }
+
+        void setOffset(uint32_t offset) { m_Offset = offset; }
+
+        uint32_t getOffset() const { return m_Offset; }
+        uint32_t getCount() const { return m_Count; }
+        uint32_t getSize() const { return m_Size; }
+        bool getNormalize() const { return m_Normalize; }
+        GLenum getType() const { return m_Type; }
+
+    private:
+        GLenum m_Type;
+        uint32_t m_Count;
+        uint32_t m_Size;
+        uint32_t m_Offset;
+        bool m_Normalize;
+    };
+
+    class BufferLayout
+    {
+    public:
+        BufferLayout() = delete;
+        ~BufferLayout() = default;
+        BufferLayout(std::initializer_list<VertexElement> vertexElements)
+        {
+            // Calculate offset for each element
+            uint32_t Offset = 0;
+            for (auto elem : vertexElements)
+            {
+                elem.setOffset(Offset);
+                m_Elements.push_back(elem);
+                Offset += elem.getSize();
+            }
+
+            m_Stride = Offset;
+        }
+
+        const std::vector<VertexElement>& getElements() const { return m_Elements; }
+        uint32_t getStride() const { return m_Stride; }
+    private:
+        std::vector<VertexElement> m_Elements;
+        uint32_t m_Stride;
+    };
+
+    class VertexArray
+    {
+    public:
+        VertexArray();
+        VertexArray(const std::shared_ptr<VertexBuffer>& vertexBuffer,
+                    const std::shared_ptr<IndexBuffer>& IndexBuffer,
```

Comment:
So I remember reading an article about when you should pass by value, reference, passing pointers (shared and unique), and wanted to look up what the conventional guidelines for this are. So obviously this constructor (plus the functions setVertexBuffer and setIndexBuffer) wants to obtain partial ownership of the underlying object (i.e increment the refcount), and here's what the [C++ core guidelines](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rr-smartptrparam) say (idk if you have read these before but they are great btw):
_"R.34: Take a shared_ptr<widget> parameter to express that a function is part owner"_
_"R.35: Take a shared_ptr<widget>& parameter to express that a function might reseat the shared pointer"_
_"R.36: Take a const shared_ptr<widget>& parameter to express that it might retain a reference count to the object ???"_
So obviously R.35 doesn't apply, because we are not reassigning the shared_ptr to another object, but I think R.34 represents our situation better than R.36 because we are always retaining a refcount when these functions/constructors are called

Reply:
TLDR: I think we might wanna change these to pass by value

Reply:
BTW this helped me better understand memory ownership, I find it very interesting haha

Reply:
Also missing our sharedPtr wrapper :(